File: src/main/java/com/example/jobs/schedule/AnalyticsRefreshScheduler.java
package com.example.jobs.schedule;

import com.example.jobs.service.AnalyticsJavaService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class AnalyticsRefreshScheduler {

  private final AnalyticsJavaService analyticsService;

  @Scheduled(cron = "0 0 * * * *")
  public void refreshAnalyticsView() {
    log.info("Triggering scheduled Materialized View refresh non-blockingly...");

    analyticsService.refreshMaterializedView()
        .doOnError(error -> log.error("Scheduled analytics refresh execution encountered an error: {}", error.getMessage(), error))
        .subscribe();
  }
}

File: src/main/kotlin/com/example/jobs/schedule/AnalyticsRefreshScheduler.kt
package com.example.jobs.schedule

import com.example.jobs.service.AnalyticsService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.scheduling.annotation.Scheduled

@Component
class AnalyticsRefreshScheduler(private val analyticsService: AnalyticsService) {

    companion object {
        private val log = LoggerFactory.getLogger(AnalyticsRefreshScheduler::class.java)
    }

    @Scheduled(cron = "0 0 * * * *")
    suspend fun refreshAnalyticsView() {
        log.info("Triggering scheduled Materialized View refresh non-blockingly via coroutine...")
        try {
            analyticsService.refreshMaterializedView()
        } catch (error: Exception) {
            log.error("Scheduled analytics refresh execution encountered an error: ${error.message}", error)
        }
    }
}